gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeUtils.java

```diff
@@ -191,24 +206,43 @@ private static boolean containsNotOrOrFilter(Filter filter) {
     return false;
   }
 
+  public static boolean areFiltersOnlyOnCurrentDataSource(ExecutionContext executionContext, String currentSource) {
+    Map<String, Set<String>> sourceToFilterAttributeMap =
+        executionContext.getSourceToFilterAttributeMap();
+
+    if (sourceToFilterAttributeMap.isEmpty()) {
+      return true;
+    }
+
+    Map<String, Set<String>> filterAttributeToSourcesMap =
+        buildAttributeToSourcesMap(sourceToFilterAttributeMap);
```

Comment:
I see this map being constructed at other places too. Does it make sense to save this within the ExecutionContext and reuse instead of computing always?

Reply:
I can do it for the filter, since that's constructed twice everytime

Reply:
done